xray/grpc.go

```diff
@@ -0,0 +1,174 @@
+			ctx2 := metadata.NewOutgoingContext(httptrace.WithClientTrace(ctx, ct.httpTrace), md)
+
+			seg.Lock()
+			seg.Namespace = "remote"
+			seg.GetHTTP().GetRequest().URL = "grpc://" + host + method
+			seg.GetHTTP().GetRequest().Method = method
+			seg.Unlock()
+
+			err := invoker(ctx2, method, req, reply, cc, opts...)
+
+			recordContentLength(seg, reply)
+			if err != nil {
+				classifyErrorStatus(seg, err)
+				ct.subsegments.GotConn(nil, err)
+			}
+
+			return err
+		})
+	}
+}
+
+func unaryServerInterceptorWithConfig(sn SegmentNamer, cfg *Config) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+		md, ok := metadata.FromIncomingContext(ctx)
+
+		var traceID string
+		if ok && len(md.Get(TraceIDHeaderKey)) == 1 {
+			traceID = md.Get(TraceIDHeaderKey)[0]
+		}
+		traceHeader := header.FromString(traceID)
+
+		var host string
+
+		if len(md.Get(":authority")) == 1 {
+			host = md.Get(":authority")[0]
+		}
+		requestURL := url.URL{
+			Scheme: "grpc",
+			Host:   host,
+			Path:   info.FullMethod,
+		}
+		name := sn.Name(host)
+
+		ctx = context.WithValue(ctx, RecorderContextKey{}, cfg)
+
+		var seg *Segment
+		ctx, seg = NewSegmentFromHeader(ctx, name, &http.Request{
+			Host:   host,
+			URL:    &requestURL,
+			Method: info.FullMethod,
+		}, traceHeader)
+		defer seg.Close(nil)
+
+		seg.Lock()
+		seg.GetHTTP().GetRequest().ClientIP, seg.GetHTTP().GetRequest().XForwardedFor = clientIPFromGrpcMetadata(md)
+		seg.GetHTTP().GetRequest().URL = requestURL.String()
+		seg.GetHTTP().GetRequest().Method = info.FullMethod
+		if len(md.Get("user-agent")) == 1 {
+			seg.GetHTTP().GetRequest().UserAgent = md.Get("user-agent")[0]
+		}
+		seg.Unlock()
+
+		resp, err = handler(ctx, req)
+		if err != nil {
+			classifyErrorStatus(seg, err)
+		}
+		recordContentLength(seg, resp)
+		if headerErr := addResponseTraceHeader(ctx, seg, traceHeader); headerErr != nil {
+			err = multierr.Combine(err, headerErr)
+		}
+
+		return resp, err
+	}
+}
+
+// UnaryServerInterceptorWithContext provides gRPC unary server interceptor. xray.Config of given ctx will be inherited
+// by the instrumented context.
+func UnaryServerInterceptorWithContext(ctx context.Context, sn SegmentNamer) grpc.UnaryServerInterceptor {
+	cfg := GetRecorder(ctx)
+	return unaryServerInterceptorWithConfig(sn, cfg)
+}
+
+// UnaryServerInterceptor provides gRPC unary server interceptor.
+func UnaryServerInterceptor(sn SegmentNamer) grpc.UnaryServerInterceptor {
+	return unaryServerInterceptorWithConfig(sn, nil)
+}
+
+func classifyErrorStatus(seg *Segment, err error) {
+	seg.Lock()
+	defer seg.Unlock()
+	grpcStatus, ok := status.FromError(err)
+	if !ok {
+		seg.Error = true
+		return
+	}
+	switch grpcStatus.Code() {
+	case codes.ResourceExhausted:
+		seg.Throttle = true
+	case codes.Internal, codes.Unimplemented, codes.DataLoss:
```

Comment:
Error is like HTTP 4xx. This list is a good one to refer to for mapping

https://github.com/open-telemetry/opentelemetry-collector/pull/3110/files#diff-3c93ea7fd05dbbc4514087f190d5c07c0d125e606559d8cdd96923081058b63dL38

Reply:
> https://github.com/open-telemetry/opentelemetry-collector/pull/3110/files#diff-3c93ea7fd05dbbc4514087f190d5c07c0d125e606559d8cdd96923081058b63dL38

The mapping doesn't classify all the possible grpc status codes here:
https://github.com/grpc/grpc-go/blob/dc77d7ffe311f78f2e577572d984af3c0a8df82b/codes/codes.go

So, I'm referencing the mapping here:
https://github.com/googleapis/googleapis/blob/b29fb6b9fcdf3bc67487fa22ab7783915023041e/google/rpc/code.proto#L35